### PR TITLE
Expand docs for `TryFrom` and `TryInto`.

### LINKED ISSUE #33417: Tracking issue for TryFrom/TryInto traits
Tracking issue for https://github.com/rust-lang/rfcs/pull/1542

----

To do:

* [ ] Is existing documentation satisfactory?
* [x] https://github.com/rust-lang/rust/pull/56796 Change bounds on `TryFrom` blanket impl to use `Into` instead of `From`
* [ ] (Re)stabilization PR

### PR DESCRIPTION
**Pull request by icefoxen**
The examples are still lacking for now, both for module docs and for methods/impl's.  Will be adding those in further pushes.

Should hopefully resolve the doc concern in #33417 when finished?

### CODE REVIEW

File: src/libcore/convert.rs [hunk 1]
```diff
@@ -376,7 +381,44 @@ pub trait TryInto<T>: Sized {
...
+/// `i32::MAX`, or by some other method.  The `From` trait is intended 
+/// for lossless conversions, so the `TryFrom` trait informs the 
+/// programmer when a type conversion could go bad and lets them
+/// decide how to handle it.
+/// 
+/// # Generic Implementations
+///
+/// - `TryFrom<T> for U` implies [`TryInto<U>`]` for T`
+/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T` 
+/// is implemented
```
Comment (scottmcm) [inline src/libcore/convert.rs]: maybe add "and cannot fail"?

Would it be worth mentioning `!` somewhere here?

Comment (icefoxen) [reply src/libcore/convert.rs]: Not sure quite what to say about `!`, maybe something like this?

```
/// - [`try_from`] is reflexive, which means that `TryFrom<T> for T`
/// is implemented and cannot fail -- the associated `Error` type
/// for calling `TryFrom<T>` on `T` is `!`
```

**Comment by Dylan-DPC** [comment]
ping from triage @icefoxen you need to resolve the failing tests.

**Comment by icefoxen** [comment]
Thanks @dpc I will gnaw on them when I get a chance.

**Comment by dpc** [comment]
Wrong dpc. @Dylan-DPC :) . But I like what's going on here, too. :tada:

**Comment by icefoxen** [comment]
There's too many dpc's around!

**Comment by icefoxen** [comment]
Ok it's time to finish pushing this through.  It currently conflicts with #58302 but the actual differences are trivial.  However the priorities work out I'll update the docs in this PR to match what we need.

**Comment by SimonSapin** [comment]
Thank you for helping push this forward!

(I’ve edited the PR description, `resolve` followed by an issue number is special on GitHub and causes that issue to be closed when the PR is merged.)

**Comment by icefoxen** [comment]
@SimonSapin Thanks!  I knew about that behavior but am never 100% sure what does and does not trigger it.  Alas.

**Comment by SimonSapin** [comment]
It’s documented in https://help.github.com/articles/closing-issues-using-keywords/. And, since… more recently than the behavior has existed, those keywords have a dotted underline and an explanation tooltip when they are interpreted specially. See for example https://github.com/rust-lang/rust/pull/58407.

**Comment by icefoxen** [comment]
Ok.  Docs done, tidy has no complaints, doctests run successfully.

r?

**Comment by Centril** [comment]
Ping from triage, @sfackler :)

File: src/libcore/num/mod.rs [hunk 2]
```diff
@@ -4449,6 +4449,9 @@ macro_rules! try_from_unbounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create the target type from the source type.
+            /// This particular variant will never fail, but is included
+            /// for completeness's sake.
```
Comment (SimonSapin) [inline src/libcore/num/mod.rs]: For some `TryFrom` conversions that involve `usize` or `isize`, details of how the conversion is implemented depend on the target platform (’s pointer size). But we typically only generate one set of documentation for the standard library, so I think we shouldn’t document details that only apply to some platforms.

The easiest way would be to have the same doc-comment for all `impl` blocks that use `TryFromIntError`.

**Comment by icefoxen** [comment]
Made the changes you asked for, or at least tried to...  it's late and I'm a little loopy, but I don't want you to have to wait on this.

**Comment by SimonSapin** [comment]
This is better, thanks. I think the word “disjoint” is not what you meant here: two ranges are disjoint if they do not overlap at all: if their intersection is empty. (And failed conversions returns `Err(..)` rather than `None`). I think we can keep it simple, something like:

> Try to create the target number type from a source number type. This returns an error if the source value is outside of the range of the target type.

**Comment by icefoxen** [comment]
Ahahaha, wow I was *really* loopy.  Should now be much better, thank you.

Do I have to rebase or anything to handle the merge conflict?  My git-fu is, as many will tell you, not great.

**Comment by SimonSapin** [comment]
Looks good!

Yes rebasing and resolving conflicts was necessary but I took care of it.

@bors r+

**Comment by SimonSapin** [comment]
This might be nice to backport to 1.34 beta since that’s when the feature was stabilized.

**Comment by icefoxen** [comment]
Huh, not sure what that test failure is.  Trying to reproduce...

File: src/libcore/num/mod.rs [hunk 3]
```diff
@@ -4559,6 +4562,9 @@ macro_rules! try_from_lower_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create the target number type from a source
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u >= 0 {
```
Comment (lzutao) [inline src/libcore/num/mod.rs]: Should this be `Result<Self, Self::Error>`?

File: src/libcore/num/mod.rs [hunk 4]
```diff
@@ -4578,6 +4584,9 @@ macro_rules! try_from_upper_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create the target number type from a source
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 if u > (<$target>::max_value() as $source) {
```
Comment (lzutao) [inline src/libcore/num/mod.rs]: here too

File: src/libcore/num/mod.rs [hunk 5]
```diff
@@ -4597,6 +4606,9 @@ macro_rules! try_from_both_bounded {
         impl TryFrom<$source> for $target {
             type Error = TryFromIntError;
 
+            /// Try to create the target number type from a source
+            /// number type. This returns an error if the source value
+            /// is outside of the range of the target type.
             #[inline]
             fn try_from(u: $source) -> Result<$target, TryFromIntError> {
                 let min = <$target>::min_value() as $source;
```
Comment (lzutao) [inline src/libcore/num/mod.rs]: and here!

**Review by tesuji — COMMENTED** [review]
This is doc PR, so these nits should be solved on another PR.

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/58799#issuecomment-468148470, @bors r-

**Comment by SimonSapin** [comment]
I took the liberty of pushing another commit to fix https://github.com/rust-lang/rust/pull/58015#issuecomment-468077048.

@bors r+

**Comment by icefoxen** [comment]
Darn it I was just working on this this morning.  XD  Sorry for the delay.

**Comment by Mark-Simulacrum** [comment]
@rust-lang/libs Can you decide if you're fine with this being backported to beta? We need a decision in the next several days so it can make it in time.

File: src/libcore/convert.rs [hunk 6]
```diff
@@ -380,7 +386,55 @@ pub trait TryInto<T>: Sized {
     fn try_into(self) -> Result<T, Self::Error>;
 }
 
-/// Attempt to construct `Self` via a conversion.
+/// Simple and safe type conversions that may fail in a controlled
+/// way under some circumstances. It is the reciprocal of [`TryInto`].
```
Comment (steveklabnik) [inline src/libcore/convert.rs]: This is a bit too much information for a summary line; they should be quite short. That said, this PR has already been merged; @icefoxen any interest in a follow-on?

**Comment by steveklabnik** [comment]
I'm okay with this being backported. Ther'es one thing that's not ideal about it but it's not a dealbreaker. :)

**Comment by alexcrichton** [comment]
The libs team discussed this today and decided it's ok to backport to beta

**Comment by Centril** [comment]
@alexcrichton Btw, when you `beta-accepted`, don't remove `beta-nominated`... it's very confusing, but it is what it is =P

File: src/libcore/convert.rs (hunk 6 above)
Comment (icefoxen) [reply src/libcore/convert.rs]: Not this time I'm afraid.  As suggested by how long it took me to notice this, life is pretty hectic at the moment.  :disappointed:

**Pull request merged**